### PR TITLE
Add quantile_uni_extrapolate preprocessor preset

### PR DESCRIPTION
**Pull request by LeoGrin**
## Summary

Adds the `quantile_uni_extrapolate` preprocessor preset: instead of clamping quantile-transformed values at the `[0, 1]` boundary, it linearly extrapolates past the boundary for inputs outside the training range, better preserving out-of-distribution information.

Wired through **both** the CPU `AdaptiveQuantileTransformer` and the GPU `TorchQuantileTransformer` paths so the preset is GPU-eligible. CPU/GPU consistency tests parametrise the new preset across f16/f32/f64.

## Motivation

v3 OOD-preprocessing checkpoints use `quantile_uni_extrapolate` in the regressor preprocessing recipe. Without this preset in the public package those checkpoints fail to load (`pydantic ValidationError` on the `PreprocessorConfig.name` literal). This unblocks loading them with the public release.

## Changes (7 source + 3 test files, mirrors private #599)

- `preprocessing/configs.py` — register `quantile_uni_extrapolate` in the name literal
- `preprocessing/steps/adaptive_quantile_transformer.py` — `extrapolate_ratio` logic (CPU)
- `preprocessing/steps/reshape_feature_distribution_step.py` — preset wiring
- `preprocessing/torch/{factory,steps,gpu_preprocessing_metadata,torch_quantile_transformer}.py` — GPU path
- tests for CPU transform, torch transform, and CPU/GPU consistency

## Test plan

- `test_adaptive_quantile_transformer.py` — 5/5 pass (incl. NaN-column & constant-feature edge cases)
- `test_torch_quantile_transformer.py` extrapolation suite — 3/3 pass (incl. `matches_cpu_on_out_of_range_inputs`)

Note: repo policy is "open an issue first" — happy to file/link one; opening this so the diff is reviewable.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes numeric preprocessing for a new checkpoint-critical preset and must keep CPU/GPU paths aligned; mistakes could shift model inputs at inference, though scope is limited to quantile uniform extrapolation with tests guarding parity.
> 
> **Overview**
> Adds **`quantile_uni_extrapolate`** as a registered `PreprocessorConfig` name so v3 OOD regressor checkpoints can load without validation errors.
> 
> The preset uses uniform quantile mapping with an optional **`extrapolate_ratio`** (default **1.0** for this preset): values outside the training min/max are linearly scaled past `[0, 1]` and clipped to `[-ratio, 1 + ratio]` instead of being flattened at the boundary, preserving more OOD signal. **CPU** `AdaptiveQuantileTransformer` stores `x_min_`/`x_max_` at fit and overrides `transform`; **GPU** `TorchQuantileTransformer` mirrors the same logic and is marked GPU-eligible. Wiring runs through reshape presets, `get_extrapolate_ratio_for_preset`, and the torch factory/selective quantile steps. Tests cover extrapolation, NaN/constant features, validation, CPU/GPU parity, and pipeline consistency (including the new preset across dtypes).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e89320abb05deeee87ae9fe68fa109c8fbcb2283. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by LeoGrin** [comment]
Thanks @gemini-code-assist — both addressed in `b54cb09`:

1. **`extrapolate_ratio` non-negative + uniform-only (CPU)** — added in `AdaptiveQuantileTransformer.__init__`: raises `ValueError` for negative ratio, and for `output_distribution != "uniform"` (linear extrapolation past `[0,1]` is only meaningful for the uniform output mapping, matching the preset's documented intent).
2. **`extrapolate_ratio` non-negative (GPU)** — added the same non-negative guard in `TorchQuantileTransformer.__init__`.

Added tests for both guard paths (`test__extrapolate_ratio__validation_guards`, `test__torch_extrapolate_ratio__rejects_negative`).

Also in this push, unrelated to the bot comments but worth flagging for review:
- **`quantile_uni_extrapolate` now uses the `n_samples // 5` tier** (same as `quantile_uni`), not the coarse `n//10` tier it had in the internal PR. Rationale: the preset is meant to be "the default quantile transform **plus** boundary extrapolation" — it should differ from `quantile_uni` by extrapolation only, not also by a coarser quantile grid. This is an intentional deviation from internal #599.
- **GPU `_apply_extrapolation` norm consolidation**: the internal port computed `norm_below` and `norm_above` as two separate full-array tensors; `(x - x_max)/range + 1` is algebraically identical to `(x - x_min)/range`, so it now computes one tensor and reuses it (the CPU side was already consolidated this way in #599). Output-identical — CPU/GPU consistency tests unchanged.

File: src/tabpfn/preprocessing/torch/factory.py [hunk 1]
```diff
@@ -66,11 +66,15 @@ def create_gpu_preprocessing_pipeline(
         quantile_on_gpu = len(quantile_target_indices) > 0
         if quantile_on_gpu and n_train_samples is not None:
             n_quantiles = compute_effective_n_quantiles(pconfig.name, n_train_samples)
+            extrapolate_ratio = (
+                1.0 if pconfig.name == "quantile_uni_extrapolate" else None
+            )
```
Comment (FBPrior) [inline src/tabpfn/preprocessing/torch/factory.py]: Do we ever want to set this to sth other than one? I think it's useful to keep this parameter around because it's a pretty important parameter.

**Review by FBPrior — APPROVED** [review]
LGTM thanks!

File: src/tabpfn/preprocessing/torch/factory.py (hunk 1 above)
Comment (LeoGrin) [reply src/tabpfn/preprocessing/torch/factory.py]: Yeah, kept it but set it to one for this preset. Is that what you have in mind?
